### PR TITLE
support loading models from json files

### PR DESCRIPTION
**Pull request by nlhepler**
- introduce model forms for each of Marginal, Snr, PwSnrA, and PwSnr
- introduce automatic counterWeight calculation for these model forms
- tests for these model forms, including loading params concordant with- and tests for correctness against their compiled analogs
- fix integrator tests to use realistic pulse widths
- fix compiled models to use computed counterWeight, restoring performance lost when using realistic pulse widths

### CODE REVIEW

**Comment by dalexander** [comment]
I would prefer the  counterweight computation to be a separate commit if not a separate PR entirely.  While it's a good direction, It also introduces new possible sources of errors.

**Comment by dalexander** [comment]
More detailed review forthcoming

**Comment by nlhepler** [comment]
Then I would make this branch depend on the counterweight PR. Because I don't want to introduce model files only to immediately change them. In any case, the test *fixes* *break* the current tests because the counterWeights were inappropriate.

File: circle.yml [hunk 1]
```diff
@@ -19,8 +19,8 @@ dependencies:
         - mkdir _rev_deps  # Create a directory for reverse-dependencies, ie things that depend on us. These are not static, do not cache them.
         - pushd _rev_deps ; git clone https://github.com/PacificBiosciences/htslib.git
         - pushd _rev_deps ; git clone https://github.com/PacificBiosciences/pbbam.git
-        - pushd _rev_deps ; git clone https://github.com/PacificBiosciences/pbccs.git
-        - pushd _rev_deps ; git clone https://${TOKEN}@github.com/PacificBiosciences/pblaa.git
+        - pushd _rev_deps ; git clone -b loadModels https://github.com/PacificBiosciences/pbccs.git
+        - pushd _rev_deps ; git clone -b loadModels https://${TOKEN}@github.com/PacificBiosciences/pblaa.git
```
Comment (dalexander) [inline circle.yml]: How can we merge this to master when it hardcodes a different branch name?  The CCS problem goes away with unanimity, but not LAA.  Are you going to come back in a subsequent commit and repoint these at master?  A bit ugly, I'd really prefer if we could figure out how to get the Circle CI variable substitution working the way we want.

Comment (dalexander) [inline circle.yml]: that said, we don't have to solve that problem today.

File: src/JsonHelpers.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2016, Pacific Biosciences of California, Inc.
```
Comment (dalexander) [inline src/JsonHelpers.h]: Why is this impl code in .h and not .cpp?  We don't need these inlined.

File: src/Recursor.h [hunk 3]
```diff
@@ -236,23 +236,23 @@ void Recursor<Derived>::FillAlpha(const M& guide, M& alpha) const
               ***********  EDGE_CONDITION ************
              */
             if (i > 0 && j > 0) {
-                thisMoveScore =
-                    alpha(i - 1, j - 1) * prevTransProbs.Match *
-                    Derived::EmissionPr(MoveType::MATCH, curReadEm, prevTplBase, currTplBase);
+                thisMoveScore = alpha(i - 1, j - 1) * prevTransProbs.Match *
```
Comment (dalexander) [inline src/Recursor.h]: what is purpose of the changes in this file?  Is it related to the model loading?  If not, please put this in a separate commit/PR.

File: src/matrix/ScaledMatrix.h [hunk 4]
```diff
@@ -69,7 +69,7 @@ class ScaledMatrix : public SparseMatrix
 
 public:  // information about entries filled by column
     template <bool maxProvided>
```
Comment (dalexander) [inline src/matrix/ScaledMatrix.h]: Style guide formatting (naming), also off-target...

File: src/matrix/ScaledMatrix.h [hunk 5]
```diff
@@ -69,7 +69,7 @@ class ScaledMatrix : public SparseMatrix
...
-    void FinishEditingColumn(int j, int usedBegin, int usedEnd, double max_val = 0.0);
+    void FinishEditingColumn(int j, int usedBegin, int usedEnd, double maxVal = 0.0);
```
Comment (dalexander) [inline src/matrix/ScaledMatrix.h]: Style guide formatting (naming), also off-target...

File: src/models/S_P1C1Beta_Model.cpp [hunk 6]
```diff
@@ -137,8 +137,8 @@ std::vector<TemplatePosition> S_P1C1Beta_Model::Populate(const std::string& tpl)
     for (size_t i = 1; i < tpl.size(); ++i) {
         const uint8_t curr = detail::TranslationTable[static_cast<uint8_t>(tpl[i])];
         if (curr > 3) throw std::invalid_argument("invalid character in sequence!");
-        const bool hpAdd = tpl[i - 1] == tpl[i] ? 0 : 4;
```
Comment (dalexander) [inline src/models/S_P1C1Beta_Model.cpp]: This was a major bug fix.  It really belongs under a separate commit, with a clear commit message indicating the fix.

File: src/models/S_P1C1Beta_Model.cpp [hunk 7]
```diff
@@ -164,7 +164,7 @@ std::unique_ptr<AbstractRecursor> S_P1C1Beta_Model::CreateRecursor(
 double S_P1C1Beta_Model::ExpectedLLForEmission(const MoveType move, const uint8_t prev,
                                                const uint8_t curr, const MomentType moment) const
 {
-    const uint8_t hpAdd = prev == curr ? 0 : 4;
```
Comment (dalexander) [inline src/models/S_P1C1Beta_Model.cpp]: IBID...

File: src/models/S_P1C1Beta_Model.cpp [hunk 8]
```diff
@@ -202,7 +202,7 @@ double S_P1C1Beta_Recursor::EmissionPr(MoveType move, const uint8_t emission, co
                                        const uint8_t curr)
 {
     assert(move != MoveType::DELETION);
-    const uint8_t hpAdd = prev == curr ? 0 : 4;
```
Comment (dalexander) [inline src/models/S_P1C1Beta_Model.cpp]: IBID

File: src/JsonHelpers.h (hunk 2 above)
Comment (nlhepler) [inline src/JsonHelpers.h]: They are templated.

File: src/Recursor.h (hunk 3 above)
Comment (nlhepler) [inline src/Recursor.h]: It is related. This is required to support EmissionPr as a method rather than static function. See https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern

File: src/matrix/ScaledMatrix.h (hunk 5 above)
Comment (nlhepler) [inline src/matrix/ScaledMatrix.h]: sure.

File: src/models/S_P1C1Beta_Model.cpp (hunk 6 above)
Comment (nlhepler) [inline src/models/S_P1C1Beta_Model.cpp]: agreed.

File: src/models/S_P1C1Beta_Model.cpp (hunk 7 above)
Comment (nlhepler) [inline src/models/S_P1C1Beta_Model.cpp]: agreed.

File: src/models/S_P1C1Beta_Model.cpp (hunk 8 above)
Comment (nlhepler) [inline src/models/S_P1C1Beta_Model.cpp]: agreed.

File: src/models/S_P1C1v1_Model.cpp [hunk 9]
```diff
@@ -272,14 +277,8 @@ S_P1C1v1_Model::S_P1C1v1_Model(const SNR& snr) : snr_(snr)
...
 
-    double snr1 = snr_.A;
     // Adjust as necessary to clamp to bounds of SNRs observe during training
-    if (snr1 < 4.0) {
-        snr1 = 4.0;
-    } else if (snr1 > 10.65) {
-        snr1 = 10.65;
-    }
-
+    const double snr1 = std::max(4.0, std::min(10.65, snr_.A));
```
Comment (dalexander) [inline src/models/S_P1C1v1_Model.cpp]: Logic seems correct here, but in the future I will consider this kind of change "off-target"

File: src/models/S_P1C1v1_Model.cpp [hunk 10]
```diff
@@ -293,23 +292,69 @@ S_P1C1v1_Model::S_P1C1v1_Model(const SNR& snr) : snr_(snr)
         }
         for (size_t j = 0; j < 4; ++j)
             ctxTrans_[ctx][j] /= sum;
-    }
 
-    // Generate cached emission expectations
-    // TODO: These are identical for all instances, either we should enrich the model or avoid doing
-    // this in a context dependent way
-    for (size_t ctx = 0; ctx < CONTEXT_NUMBER; ++ctx)
+        // Generate cached emission expectations
```
Comment (dalexander) [inline src/models/S_P1C1v1_Model.cpp]: off-target whitespace formatting ...

File: src/JsonHelpers.h (hunk 2 above)
Comment (dalexander) [inline src/JsonHelpers.h]: Ah, yes.

File: src/models/S_P1C1v1_Model.cpp (hunk 9 above)
Comment (nlhepler) [inline src/models/S_P1C1v1_Model.cpp]: let me just remove it.

File: src/Recursor.h (hunk 3 above)
Comment (dalexander) [inline src/Recursor.h]: OK thanks

File: src/Recursor.h [hunk 11]
```diff
@@ -499,7 +500,7 @@ void Recursor<Derived>::ExtendAlpha(const M& alpha, size_t beginColumn, M& ext,
 
         int i;
         double score = 0.0;
-        double max_score = score;
+        double maxScore = score;
```
Comment (dalexander) [inline src/Recursor.h]: off-target

File: src/Recursor.h [hunk 12]
```diff
@@ -629,7 +632,7 @@ void Recursor<Derived>::ExtendBeta(const M& beta, size_t lastColumn, M& ext, int
 
         TemplatePosition currTplParams = kDefaultTplPos;
         if (jp > 0) currTplParams = (*tpl_)[jp - 1];
-        double max_score = 0.0;
```
Comment (dalexander) [inline src/Recursor.h]: off-target

File: tests/TestLoadModels.cpp [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+    "CTGCGCCGGGAGGTTGAAGAACTGCGGCAGGCCAGCGAGGCAGATCTCCAGCCAGGAACTATTGAGTACGAACG"
+    "CCATCGACTTACGCGTGCGCAGGCCGACGCACAGGAACTGAAGAATGCCAGAGACTCCGCTGAAGTGGTGGAAA"
+    "CCGCATTCCCCTGTACTTTCGTGCTGTCGCGGATCGCAGGTGAAATTGCCAGTATTCTCGACGGGCTCCCCCTG"
+    "TCGGTGCAGCGGCGTTTTCCGGAACTGGAAAACCGACATGTTGATTTCCTGAAACGGGATATCATCAAAGCCAT"
+    "GAACAAAGCAGCCGCGCTGGATGAACTGATACCGGGGTTGCTGAGTGAATATATCGAACAGTCAGGTTAACAGG"
+    "CTGCGGCATTTTGTCCGCGCCGGGCTTCGCTCACTGTTCAGGCCGGAGCCACAGACCGCCGTTGAACGGATGCT"
+    "AATTACTATCTCCCGAAAGAATC";
+
+const std::vector<uint8_t> longPws(837, 2);
+const IntegratorConfig cfg(std::numeric_limits<double>::quiet_NaN());
```
Comment (dalexander) [inline tests/TestLoadModels.cpp]: There's no documentation on `IntegratorConfig`.  I presume the intent here is that passing NaN as the min zscore arg disables zscore filtering?

File: src/JsonHelpers.h (hunk 2 above)
Comment (nlhepler) [inline src/JsonHelpers.h]: It's because the matrices passed in have static sizes.

File: tests/TestLoadModels.cpp [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+
+TEST(LoadModelsTest, Directory)
+{
+    ASSERT_TRUE(LoadModels(tests::DataDir + "/params"));
+    std::set<std::string> chems = SupportedModels();
+    ASSERT_TRUE(chems.find("S/P1-C1/beta::Marginal") != chems.end());
+    ASSERT_TRUE(chems.find("S/P1-C1.1::PwSnrA") != chems.end());
+    ASSERT_TRUE(chems.find("S/P1-C1.2::PwSnr") != chems.end());
+
+    // test identity between S/P1-C1/beta and S/P1-C1/beta::Marginal (loaded)
```
Comment (dalexander) [inline tests/TestLoadModels.cpp]: It seems weird to me that you are overloading the nomenclature thusly:

chemistry::{modelform}

or

chemistry::Compiled

Is it never going to be the case that we compile in more than one model form for a chemistry?  I think we can live with this for this PR but it does raise questions for the future.  One might want to consider something like

Chemistry::ModelForm[::{Compiled,LoadedFromFile}]

where perhaps if the last entity is omitted, we try the file, then compiled.

Comment (nlhepler) [inline tests/TestLoadModels.cpp]: We might as well get it right this time.

File: tests/TestLoadModels.cpp [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+TEST(LoadModelsTest, ModelOverride)
+{
+    MonoMolecularIntegrator ai1(longTpl, cfg, snr, "S/P1-C1.2");
+    EXPECT_EQ(State::VALID,
+              ai1.AddRead(MappedRead(MkRead(longRead, snr, "S/P1-C1.2", longPws),
+                                     StrandType::FORWARD, 0, longTpl.length(), true, true)));
+
+    ASSERT_TRUE(OverrideModel("S/P1-C1.2"));
+
+    MonoMolecularIntegrator ai2(longTpl, cfg, snr, "S/P1-C1.1");
```
Comment (dalexander) [inline tests/TestLoadModels.cpp]: (unrelated) question: why is the chemistry being declared to both the MMI and the MappedRead?  This seems like a wart

File: tests/TestIntegrator.cpp [hunk 16]
```diff
@@ -107,38 +107,49 @@ const string longRead =
     "AATTACTATCTCCCGAAAGAATC";
 const IntegratorConfig cfg(std::numeric_limits<double>::quiet_NaN());
 
-Read MkRead(const string& seq, const SNR& snr, const string& mdl, const vector<uint8_t>& pw)
+Read MkRead(const string& seq, const SNR& snr, const string& mdl, const vector<uint8_t>& pws)
 {
     vector<uint8_t> ipd(0, seq.length());
```
Comment (dalexander) [inline tests/TestIntegrator.cpp]: This is an incorrect use of the vector constructor.  The first argument should be the length, the second is the value.

File: tests/TestLoadModels.cpp (hunk 15 above)
Comment (nlhepler) [inline tests/TestLoadModels.cpp]: MonoMolecularIntegrator requires that each read be of the same chemistry. We actually test this and throw an exception otherwise.

File: tests/TestIntegrator.cpp (hunk 16 above)
Comment (nlhepler) [inline tests/TestIntegrator.cpp]: That belongs in a separate commit, like the varname changes.

Comment (nlhepler) [inline tests/TestIntegrator.cpp]: In fact, all these PW-related changes.

File: tests/TestIntegrator.cpp [hunk 17]
```diff
@@ -107,38 +107,49 @@ const string longRead =
...
-    return Read("NA", seq, ipd, pw, snr, mdl);
+    return Read("NA", seq, ipd, pws, snr, mdl);
 }
 
 #if EXTENSIVE_TESTING
```
Comment (dalexander) [inline tests/TestIntegrator.cpp]: Unless I am mistaken, the changes in this file have nothing to do with the model loading and everything to do with the counterweights.  Could you remove these file changes from this commit/PR and stage them in your follow-on counterweight-fixing commit?

File: tests/RandomDNA.cpp [hunk 18]
```diff
@@ -33,10 +33,25 @@
 // OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 // SUCH DAMAGE.
 
+#include <iterator>
+
 #include "RandomDNA.h"
 
 using namespace std;
```
Comment (dalexander) [inline tests/RandomDNA.cpp]: This is related to the counterweight changes, not the model loading

Comment (nlhepler) [inline tests/RandomDNA.cpp]: You're right.

**Comment by dalexander** [comment]
Some "overall" comments:

- there are some very off-target diffs in here, for example the fix of the bool/int8 bug you found last week absolutely needs to be called out as a separate commit.  

- I really don't like that you entangled the counterweights work with the model loading work---I've agreed to tolerate it since you've disabled the mechanism via "ifdef 0", but I'm unhappy that there are still huge diffs in here (in the testing and RandomDNA code, for example)  that have no relation to the model loading work.  I can tolerate this for this PR only, going forward please keep things tidier.

- We've built up a substantial documentation deficit and it only gets worse with this commit.  There are not even comments describing what each model form is, or how the different Model* types relate to one another.  I'd like that to be improved before we approve this.

File: tests/TestLoadModels.cpp (hunk 15 above)
Comment (dalexander) [inline tests/TestLoadModels.cpp]: OK that seems good then

File: src/ModelFactory.cpp [hunk 19]
```diff
@@ -44,40 +49,57 @@
...
-}
-
 std::unique_ptr<ModelConfig> ModelFactory::Create(const std::string& name, const SNR& snr)
 {
-    const auto it = CreatorTable().find(name);
+    std::string model(name);
+
+    if (ModelOverride())
+        model = *ModelOverride();
+    else if (name.find("::") == std::string::npos) {
```
Comment (dalexander) [inline src/ModelFactory.cpp]: A comment would be nice here!  This, IIUC, is the mechanism for choosing the model form when only the chemistry is specified.  And there is some policy happening here about which models are best.

File: include/pacbio/consensus/ModelSelection.h [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+
+#include <boost/none.hpp>
+
+namespace PacBio {
+namespace Consensus {
+
+std::set<std::string> SupportedModels();
+std::set<std::string> SupportedChemistries();
+
+bool OverrideModel(const boost::none_t& none);
```
Comment (dalexander) [inline include/pacbio/consensus/ModelSelection.h]: Do we need this none_t business?  How about just OverrideModel() or UnOverrideModel()

File: src/JsonHelpers.h (hunk 2 above)
Comment (dalexander) [inline src/JsonHelpers.h]: Ah, I see that now.  Good enough.

File: src/ModelFormFactory.h [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+        return std::unique_ptr<ModelCreator>(new T(pt));
+    }
+};
+
+#define REGISTER_MODELFORM(cls) \
+private:                        \
+    static const ModelFormCreatorImpl<cls> creator_
+
+#define REGISTER_MODELFORM_IMPL(cls) const ModelFormCreatorImpl<cls> cls::creator_(cls::Name())
+
```
Comment (dalexander) [inline src/ModelFormFactory.h]: I will review this design pattern at a later time, since we already use it.

File: src/ModelSelection.cpp [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+bool OverrideModel(const boost::none_t& none)
+{
+    ModelOverride() = boost::none;
+    return true;
+}
+
+bool OverrideModel(const std::string& name)
+{
+    std::set<std::string> models = SupportedModels();
+    if (name.find("::") != std::string::npos) {
```
Comment (dalexander) [inline src/ModelSelection.cpp]: This logic is duplicated in `ModelFactory::Create`.  Can we consolidate it in one place?

File: src/models/P6C4NoCovModel.cpp [hunk 23]
```diff
@@ -49,14 +51,16 @@ namespace Consensus {
 namespace {
 
 constexpr double kEps = 0.00505052456472967;
```
Comment (dalexander) [inline src/models/P6C4NoCovModel.cpp]: off-topic: what is `kEps`, and how was this value chosen?

File: src/models/P6C4NoCovModel.cpp [hunk 24]
```diff
@@ -125,6 +135,65 @@ double P6C4NoCovParams[4][2][3][4] = {
 P6C4NoCovModel::P6C4NoCovModel(const SNR& snr)
     : snr_(ClampSNR(snr, SNR(0, 0, 0, 0), SNR(20, 19, 20, 20)))
 {
+    for (size_t ctx = 0; ctx < CONTEXT_NUMBER; ++ctx) {
```
Comment (dalexander) [inline src/models/P6C4NoCovModel.cpp]: Code like this is what worries me most---it's new code and somewhat algorithmic.  I'm not sure how confidently I can review this.  Is there some existing code I can compare to?

Comment (nlhepler) [inline src/models/P6C4NoCovModel.cpp]: This should be delayed till the counterWeight auto-computation. It's present in every other model +SNR model, but it should wait.

File: src/models/P6C4NoCovModel.cpp (hunk 23 above)
Comment (nlhepler) [inline src/models/P6C4NoCovModel.cpp]: It's the substitutionRate parameter

Comment (nlhepler) [inline src/models/P6C4NoCovModel.cpp]: It was trained.

File: tests/TestLoadModels.cpp (hunk 15 above)
Comment (nlhepler) [inline tests/TestLoadModels.cpp]: We could change the API to do so.

Comment (nlhepler) [inline tests/TestLoadModels.cpp]: In a different commit.

**Comment by nlhepler** [comment]
In response to overall comments:

1) I agree, it should be a separate commit.
2) Yea, I let too much leaked through. I'm going to excise all of those changes, in the spirit of getting this one right.
3) I can add a design document and fill out these new sections.

Also, for future reference: let's not do "let it slip this time." Now, or ever. If I'm getting called out for poor practices, I don't want special treatment in the subsequent breath. If this is going to be reviewed according to the new practices then we might as well take the time to follow them. Anything else is professionally embarrassing.

File: circle.yml [hunk 25]
```diff
@@ -47,7 +47,7 @@ test:
         - pushd _build ; make check
         - pushd _build ; rm -rf * ; CC=gcc-4.8 CXX=g++-4.8 $(readlink -f ../_deps/cmake-3.3.0-Linux-x86_64/bin/cmake) -DCMAKE_BUILD_TYPE=Release -DBoost_INCLUDE_DIRS=$(readlink -f ../_deps/boost_1_58_0) -DEXTENSIVE_TESTING=1 ..
         - pushd _build ; make clean ; make test_pbconsensus
-        - tests/bin/test_pbconsensus --gtest_filter=IntegratorTest.TestLongTemplateTiming*
+        - tests/bin/test_pbconsensus --gtest_filter=IntegratorTest.TestLongTemplateTiming*:LoadModelsTest.ModelTiming
```
Comment (dalexander) [inline circle.yml]: are these the only CC2 tests being run?  why aren't we running the rest?

File: src/models/P6C4NoCovModel.cpp (hunk 23 above)
Comment (dalexander) [inline src/models/P6C4NoCovModel.cpp]: OK,  that's fine.  It's an odd choice of name since I would have expected eps to have something to do with machine precision.

Comment (evolvedmicrobe) [inline src/models/P6C4NoCovModel.cpp]: It came from the practice of representing the probability-of-an-error/error-rate with the epsilon symbol (same first letter and all), e.g. (1-e) * Blah + e * Blah type things.

It's no longer used in the newer models though.

**Comment by dalexander** [comment]
In response to your comment on my summary, I was only offering the option of "letting it slip this time" because I don't want to be authoritarian on these issues because I only broadcast my policy on such matters *earlier today*---I was not offering *you* special treatment, I was offering *this PR( the treatment because I was considering it granfathered in.  

But if you are willing to roll up your sleeves and get it right in this PR, I am all in favor.

File: src/ModelSelection.cpp (hunk 22 above)
Comment (nlhepler) [inline src/ModelSelection.cpp]: Yep!

**Comment by nlhepler** [comment]
@dalexander assuming this passes its checks, it should be ready for another round of review. In the meantime I'll work on getting the other changes (fixing PW in tests, counterWeight auto-calc, broken S/P1-C1/beta model, code-gardening) into other PRs.

**Comment by dalexander** [comment]
Dear Lance,  thank you for this more targeted commit, it is indeed "surgically precise" as you said---kudos!  I like the way you cleaned up the string tokenizing.  And overall I think the design seems quite sound! A few generic questions/comments:

- Would you mind adding a single line comment for each of the new classes (`ModelFormFactory`,  `ModelFactory`, `CreatorTable` etc.) describing the role that each plays?  I have a little trouble knowing how they relate and what each does.  Ultimately I can make a picture of this for the design doc, but I need your help to understand this "cast of characters".

- How can we review or be confident in the new model code?  The new "ModelForm" classes that you have written are all entirely new--how can we be sure there are no subtle coding errors?

Another question that I don't think we can address in this PR, but which I want to raise and for you and Armin to think about: in the future how can we consolidate the "ModelForm" and "Model" implementations?  I understand we want to use the hard-baked "Model" class for runtime benefits (belief: inlining is eliding code paths related to structural zeros in the model parameters), but is there a way we can achieve this goal without duplicating implementations?  I don't know the answer ATM.

**Comment by nlhepler** [comment]
Sure! I'll do this as a follow-up commit to the same PR.

WRT to modelform confidence: The new model code, at least the PwSnrA and PwSnr models, are tested against their compiled-in counterparts (S/P1-C1v1 and S/P1-C1v2 respectively). And so will the Marginal model when I fix the S/P1-C1/beta model. The Snr model form requires a P6-C4 model parameterization that is fixed off match instead of branch, but I imagine we'll have one of those eventually.

**Comment by nlhepler** [comment]
WRT to your final statement, I wish this as well: there is a lot of duplicated code at present. I think Armin and I will need to sit down and think about which parts can be shared and which cannot, and think about how a redesign might share code. I think this could probably take 4 days or so at most.

**Comment by nlhepler** [comment]
@dalexander: ModelFactory et al are not introduced via this commit. Do you want comments there as well?

**Comment by dalexander** [comment]
Thanks for the comments.   I am OK with this now, and I agree we should sit with Armin and figure out how to either 
- consolidate the two model implementations, 
- or whether we could even eliminate the original implementation (runtime benefit of 7% is good but not necessarily worth the problems we could run into from duplication) in favor of just the new impl (which we could still make work with compiled-in parameters, though maybe without a performance benefit from inlining.)
